src/migration.types.ts

```diff
@@ -75,22 +77,77 @@ export interface ErrorReport {
 }
 
 /**
- * Callback containing the logic to migrate a dashboard from one version to another.
+ * A function that can be called to migrate a dashboard from one version to another.
  */
 export type MigrateDashboardCallback<FromDashboardState, ToDashboardState> = (
   dashboardState: FromDashboardState,
-) => ToDashboardState;
+  {
+    dataModels,
+    keysOfWidgetPluginsToRemove,
+  }: {
+    dataModels: { [serverKey: string]: DataModel };
+    keysOfWidgetPluginsToRemove: string[];
+  },
+) => void | ToDashboardState;
```

Comment:
Why the union of the return types?

Reply:
To align on the types of the recipe `produce`: you can either mutate, or return a completely new object.
I think it makes sense even if the caller doesn't know that it will ultimately be used in `produce`.